src/download.rs

```diff
@@ -58,35 +60,71 @@ impl SnapshotDownloader {
     ) -> Result<()> {
         let path = path.as_ref();
         let _ = path.file_name().context(error::ValidateFileName { path })?;
-        let target_dir = path
-            .parent()
-            .context(error::ValidateParentDirectory { path })?;
 
         // Find the overall volume size, the block size, and the metadata we need for each block:
         // the index, which lets us calculate the offset into the volume; and the token, which we
         // need to retrieve it.
-        let Snapshot {
-            volume_size,
-            block_size,
-            blocks,
-        } = self.list_snapshot_blocks(snapshot_id).await?;
-
-        // Create a temporary file and extend it to the required size.
-        let temp_file = NamedTempFile::new_in(target_dir)
-            .context(error::CreateTempFile { path: target_dir })?;
-        let temp_file_len = volume_size * GIBIBYTE;
-        let temp_file_len = u64::try_from(temp_file_len).with_context(|| error::ConvertNumber {
-            what: "temp file length",
-            number: temp_file_len.to_string(),
-            target: "u64",
-        })?;
-        temp_file
-            .as_file()
-            .set_len(temp_file_len)
-            .context(error::ExtendTempFile {
-                path: temp_file.as_ref(),
+        let snapshot: Snapshot = self.list_snapshot_blocks(snapshot_id).await?;
+
+        let mut is_block_device: bool = false;
+        if path.exists() {
+            let file_meta = fs::metadata(path)
+                .await
+                .context(error::ReadFileMetadata { path })?;
```

Comment:
I'd prefer some sort of abstraction around the two different targets - something like [this](https://gist.github.com/bcressey/fe9a26074e417e3fe65f4199d92ee768) that uses a trait to express the common interface.

Then the function here could be:
```
let target = if BlockDeviceTarget::is_valid(path)? {
    BlockDeviceTarget::new(path)?
} else if FileTarget::is_valid(path)? {
    FileTarget::new(path)?
} else {
    // error
};

target.grow(snapshot.volume_size * GIBIBYTE)?;
self.write_snapshot_blocks(&snapshot_id, snapshot, &target, progress_bar).await?;
target.finalize()?;
```

Reply:
Thanks Ben, for the feedback and reviewing. I will make the changes along the lines you suggested.

Reply:
Thanks Ben, for the feedback. I have made the changes.